File: pkg/templates/typescript/moondream-groq-computer-use/moondream.ts
export interface MoondreamPoint {
  x: number;
  y: number;
}

export class MoondreamError extends Error {}

export class MoondreamClient {
  private apiKey: string;
  private baseUrl: string;
  private timeoutMs: number;

  constructor({
    apiKey,
    baseUrl = 'https://api.moondream.ai/v1',
    timeoutMs = 30000,
  }: {
    apiKey: string;
    baseUrl?: string;
    timeoutMs?: number;
  }) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.timeoutMs = timeoutMs;
  }

  async query(imageBase64: string, question: string, reasoning?: boolean): Promise<string> {
    const payload: Record<string, unknown> = {
      image_url: toDataUrl(imageBase64),
      question,
    };
    if (reasoning !== undefined) payload.reasoning = reasoning;
    const data = await this.post('/query', payload);
    if (typeof data.answer !== 'string') {
      throw new MoondreamError('Moondream query returned an invalid response');
    }
    return data.answer;
  }

  async caption(imageBase64: string, length: 'short' | 'normal' | 'long' = 'normal'): Promise<string> {
    const payload = {
      image_url: toDataUrl(imageBase64),
      length,
      stream: false,
    };
    const data = await this.post('/caption', payload);
    if (typeof data.caption !== 'string') {
      throw new MoondreamError('Moondream caption returned an invalid response');
    }
    return data.caption;
  }

  async point(imageBase64: string, objectLabel: string): Promise<MoondreamPoint | null> {
    const payload = {
      image_url: toDataUrl(imageBase64),
      object: objectLabel,
    };
    const data = await this.post('/point', payload);
    if (!Array.isArray(data.points) || data.points.length === 0) {
      return null;
    }
    const point = data.points[0] as { x?: number; y?: number };
    if (typeof point?.x !== 'number' || typeof point?.y !== 'number') {
      return null;
    }
    return { x: point.x, y: point.y };
  }

  async detect(
    imageBase64: string,
    objectLabel: string,
  ): Promise<Array<{ x_min: number; y_min: number; x_max: number; y_max: number }>> {
    const payload = {
      image_url: toDataUrl(imageBase64),
      object: objectLabel,
    };
    const data = await this.post('/detect', payload);
    if (!Array.isArray(data.objects)) {
      return [];
    }
    const results: Array<{ x_min: number; y_min: number; x_max: number; y_max: number }> = [];
    for (const item of data.objects) {
      const box = item as { x_min?: number; y_min?: number; x_max?: number; y_max?: number };
      if ([box.x_min, box.y_min, box.x_max, box.y_max].every(v => typeof v === 'number')) {
        results.push({
          x_min: box.x_min as number,
          y_min: box.y_min as number,
          x_max: box.x_max as number,
          y_max: box.y_max as number,
        });
      }
    }
    return results;
  }

  private async post(path: string, payload: Record<string, unknown>): Promise<Record<string, any>> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Moondream-Auth': this.apiKey,
        },
        body: JSON.stringify(payload),
        signal: controller.signal,
      });

      if (!response.ok) {
        const text = await response.text();
        throw new MoondreamError(`Moondream API error ${response.status}: ${text}`);
      }

      const data = await response.json();
      if (!data || typeof data !== 'object') {
        throw new MoondreamError('Moondream API returned unexpected response type');
      }
      return data as Record<string, any>;
    } catch (error) {
      if (error instanceof MoondreamError) throw error;
      if (error instanceof Error) {
        throw new MoondreamError(error.message);
      }
      throw new MoondreamError(String(error));
    } finally {
      clearTimeout(timeout);
    }
  }
}

function toDataUrl(imageBase64: string): string {
  return `data:image/png;base64,${imageBase64}`;
}

File: pkg/templates/python/moondream-groq-computer-use/moondream.py
"""Async client for the Moondream API."""

from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any, Optional

import httpx


@dataclass
class MoondreamPoint:
    x: float
    y: float


class MoondreamError(RuntimeError):
    pass


class MoondreamClient:
    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.moondream.ai/v1",
        timeout: float = 30.0,
    ) -> None:
        self._client = httpx.AsyncClient(
            base_url=base_url,
            timeout=timeout,
            headers={
                "Content-Type": "application/json",
                "X-Moondream-Auth": api_key,
            },
        )

    async def __aenter__(self) -> "MoondreamClient":
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.close()

    async def close(self) -> None:
        await self._client.aclose()

    async def query(self, image_base64: str, question: str, reasoning: Optional[bool] = None) -> str:
        payload: dict[str, Any] = {
            "image_url": _to_data_url(image_base64),
            "question": question,
        }
        if reasoning is not None:
            payload["reasoning"] = reasoning

        data = await self._post("/query", payload)
        answer = data.get("answer")
        if not isinstance(answer, str):
            raise MoondreamError("Moondream query returned an invalid response")
        return answer

    async def caption(self, image_base64: str, length: str = "normal") -> str:
        payload = {
            "image_url": _to_data_url(image_base64),
            "length": length,
            "stream": False,
        }
        data = await self._post("/caption", payload)
        caption = data.get("caption")
        if not isinstance(caption, str):
            raise MoondreamError("Moondream caption returned an invalid response")
        return caption

    async def point(self, image_base64: str, obj: str) -> Optional[MoondreamPoint]:
        payload = {
            "image_url": _to_data_url(image_base64),
            "object": obj,
        }
        data = await self._post("/point", payload)
        points = data.get("points")
        if not isinstance(points, list) or not points:
            return None
        point = points[0]
        if not isinstance(point, dict):
            return None
        x = point.get("x")
        y = point.get("y")
        if not isinstance(x, (int, float)) or not isinstance(y, (int, float)):
            return None
        return MoondreamPoint(x=float(x), y=float(y))

    async def detect(self, image_base64: str, obj: str) -> list[dict[str, float]]:
        payload = {
            "image_url": _to_data_url(image_base64),
            "object": obj,
        }
        data = await self._post("/detect", payload)
        objects = data.get("objects")
        if not isinstance(objects, list):
            return []
        results: list[dict[str, float]] = []
        for item in objects:
            if not isinstance(item, dict):
                continue
            x_min = item.get("x_min")
            y_min = item.get("y_min")
            x_max = item.get("x_max")
            y_max = item.get("y_max")
            if all(isinstance(v, (int, float)) for v in (x_min, y_min, x_max, y_max)):
                results.append(
                    {
                        "x_min": float(x_min),
                        "y_min": float(y_min),
                        "x_max": float(x_max),
                        "y_max": float(y_max),
                    }
                )
        return results

    async def _post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        response = await self._client.post(path, json=payload)
        if response.status_code >= 400:
            text = response.text
            raise MoondreamError(f"Moondream API error {response.status_code}: {text}")
        data = response.json()
        if not isinstance(data, dict):
            raise MoondreamError("Moondream API returned unexpected response type")
        return data


def _to_data_url(image_base64: str) -> str:
    # Kernel screenshots are PNG by default
    return f"data:image/png;base64,{image_base64}"


def encode_image_bytes(image_bytes: bytes) -> str:
    return base64.b64encode(image_bytes).decode("utf-8")
